### PR TITLE
Defer support

### PR DESCRIPTION
**Pull request by laverdet**
This builds on #29 by implementing a "more correct" approach. `Readable.from` is used to convert the `asyncIterable` to a `Readable` which is then forwarded to Koa with little fuss. This implementation should have slightly better error and cancellation propagation semantics.

Also `next` is no longer invoked on error, since this was definitely not correct and prevents error middleware from working correctly. This would cause the server to respond with a status of 200 and the default body of "OK". Since GraphQL will return JSON error responses (instead of throwing) this change should only affect the code path where a multipart response was rejected due to lack of support. Therefore, a major semver change is not needed.

@matthew-gordon @trevor-scheer

### CODE REVIEW

**Comment by trevor-scheer** [comment]
Hey @laverdet - I checked out the PR and it looks like the defer integration tests are failing. This isn't obvious...they're skipped locally by default and CI isn't running against your PR.

For local dev you'll need to set the environment variable to enable them (which the workflow does) like so:
`INCREMENTAL_DELIVERY_TESTS_ENABLED=t npm test`

**Comment by laverdet** [comment]
Thank you for pointing that out. These tests are also failing in the core apollo-server package.

```
$ git clone git@github.com:apollographql/apollo-server.git
$ cd apollo-server
$ npm i -D
$ INCREMENTAL_DELIVERY_TESTS_ENABLED=thanks npm test
[...]
  ● integration tests › httpServerTests.ts › incremental delivery › tests that require graphql@17 › basic @defer working with accept: multipart/mixed; deferSpec=20220824

    expect(received).toEqual(expected) // deep equality

    Expected: 200
    Received: 406

      2350 |                 query: '{ first: testString ... @defer { testString } }',
      2351 |               });
    > 2352 |             expect(res.status).toEqual(200);
           |                                ^
      2353 |             expect(res.header['content-type']).toMatchInlineSnapshot(
      2354 |               `"multipart/mixed; boundary="-"; deferSpec=20220824"`,
      2355 |             );

      at ../integration-testsuite/src/httpServerTests.ts:2352:32
[...]
  ● integration tests › httpServerTests.ts › incremental delivery › tests that require graphql@17 › basic @defer working with accept: multipart/mixed; deferSpec=20220824, application/json

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `integration tests httpServerTests.ts incremental delivery tests that require graphql@17 basic @defer working with accept: multipart/mixed; deferSpec=20220824, application/json 1`

    Snapshot: "multipart/mixed; boundary="-"; deferSpec=20220824"
    Received: "application/json; charset=utf-8"

      2351 |               });
      2352 |             expect(res.status).toEqual(200);
    > 2353 |             expect(res.header['content-type']).toMatchInlineSnapshot(
           |                                                ^
      2354 |               `"multipart/mixed; boundary="-"; deferSpec=20220824"`,
      2355 |             );
[...]
```

and so on. The result of `executeHTTPGraphQLRequest` is not correct according to the tests but that's certainly out of the scope of this PR. #29 fails the new tests as well.

**Comment by trevor-scheer** [comment]
Oversight on my part. I'll be at my desk in a bit to try this out but you might beat me to it. My instructions also require you to install `graphql@alpha` locally for those tests to pass. Sorry about that! Yours very well may be in good shape.

**Comment by trevor-scheer** [comment]
I was able to validate that tests are passing here, sorry for the confusion!

File: src/index.ts [hunk 1]
```diff
@@ -1,11 +1,13 @@
+import { Readable } from 'node:stream';
+import { parse } from 'node:url';
 import type { WithRequired } from '@apollo/utils.withrequired';
+import type { HTTPGraphQLResponseBody } from '@apollo/server/dist/esm/externalTypes/http';
```
Comment (trevor-scheer) [inline src/index.ts:4]: We shouldn't import from `/dist`. This type isn't publicly exported (though it could be).

For now, we could access the type via `HTTPGraphQLResponse['body']`.

Though I'd argue that where we're using it, TypeScript is actually correct in that it's a `never`. Whatever we have in that `else` conditional is _not_ an `HTTPGraphQLResponseBody` and we should probably handle it as an `unknown` and just throw a generic programming error that says the body was an unexpected type.

File: src/index.ts [hunk 2]
```diff
@@ -57,51 +59,59 @@ export function koaMiddleware<TContext extends BaseContext>(
...
-        // multi-valued headers with `, ` just like the Fetch API's `Headers`
-        // does. We assume that keys are already lower-cased (as per the Node
-        // docs on IncomingMessage.headers) and so we don't bother to lower-case
-        // them or combine across multiple keys that would lower-case to the
-        // same value.
-        headers.set(
-          key,
-          Array.isArray(value) ? value.join(', ') : (value as string),
-        );
+    const incomingHeaders = new Map(function*() {
```
Comment (trevor-scheer) [inline src/index.ts:62]: Is this change necessary or beneficial? I only see additional complexity by introducing an II(generator)FE.

**Review by trevor-scheer — CHANGES_REQUESTED** [review]
Just a couple minor requests, otherwise LGTM! Thanks for putting this together!

File: src/index.ts (hunk 1 above)
Comment (laverdet) [reply src/index.ts]: Thanks, this was auto-suggested by VS Code and I didn't look at it too hard. This symbol is not exported by Apollo so I guess we'll just have to use `any` due to the enum type narrowing in the bottom `else`.

File: src/index.ts (hunk 2 above)
Comment (laverdet) [reply src/index.ts]: It's not necessary, just a personal style thing. I was changing the name of this variable and the rest was reflex. I use this IIFE style a lot because it means you don't need to specify types [`new Map` vs `new Map<string, string>`], you don't need to repeat the name of the container [`yield` vs `incomingHeaders.set(...)`], and if a temporary variable is needed then it is contained to a nested scope. This also mirrors [do expressions](https://github.com/tc39/proposal-do-expressions) which will hopefully make it into the language in one way or another. Especially with TypeScript this style is super helpful. Anyway I reverted the change because it's not important.
